src/commands.ts

```diff
@@ -8,42 +8,79 @@ import { Remote } from "./remote"
 import { Storage } from "./storage"
 import { OpenableTreeItem } from "./workspacesProvider"
 
-// maybeAskUrl asks the user for the URL if it was not provided and normalizes
-// the returned URL.
-export async function maybeAskUrl(
-  providedUrl: string | undefined | null,
-  lastUsedUrl?: string,
-): Promise<string | undefined> {
-  let url =
-    providedUrl ||
-    (await vscode.window.showInputBox({
-      title: "Coder URL",
-      prompt: "Enter the URL of your Coder deployment.",
-      placeHolder: "https://example.coder.com",
-      value: lastUsedUrl,
-    }))
-  if (!url) {
-    return undefined
-  }
-  if (!url.startsWith("http://") && !url.startsWith("https://")) {
-    // Default to HTTPS if not provided!
-    // https://github.com/coder/vscode-coder/issues/44
-    url = "https://" + url
-  }
-  while (url.endsWith("/")) {
-    url = url.substring(0, url.length - 1)
-  }
-  return url
-}
-
 export class Commands {
   public constructor(
     private readonly vscodeProposed: typeof vscode,
     private readonly storage: Storage,
   ) {}
 
+  /**
+   * Ask the user for the URL, letting them choose from a list of recent URLs or
+   * CODER_URL or enter a new one.  Undefined means the user aborted.
+   */
+  private askURL(selection?: string): Promise<string | undefined> {
+    const quickPick = vscode.window.createQuickPick()
+    quickPick.value = selection || process.env.CODER_URL || ""
+    quickPick.placeholder = "https://example.coder.com"
+    quickPick.title = "Enter the URL of your Coder deployment."
+
+    // Initial items.
+    quickPick.items = this.storage.withUrlHistory(process.env.CODER_URL).map((url) => ({
+      alwaysShow: true,
+      label: url,
+    }))
+
+    // Quick picks do not allow arbitrary values, so we add the value itself as
+    // an option in case the user wants to connect to something that is not in
+    // the list.
+    quickPick.onDidChangeValue((value) => {
```

Comment:
What happens if the user enters a duplicate or partial URL?

Reply:
`withUrlHistory` removes duplicates (using a fancy `Set`).

Reply:
Oh and a partial URL will show at the top of the list, like this:

![screenshot](https://github.com/coder/vscode-coder/assets/45609798/e822a39c-c10a-41f1-8d5c-20a5fe7b3f65)

Here is the duplicate removal at work, if you type something that matches, it only shows it once:

![screenshot](https://github.com/coder/vscode-coder/assets/45609798/c2b1fe2d-fc63-4b6a-adcf-6d320a877f93)

Reply:
One improvement we could make is that if they omit the `https://` we could consider it still the same, right now it would show twice, even though if they accept the `dev.coder.com` option we will add `https://` to it (when we save the URL history we will only save one though).

![screenshot](https://github.com/coder/vscode-coder/assets/45609798/c100d338-c63c-48fa-9fef-72f77796d3ea)

Reply:
That seems reasonable; what you have right now isn't offensive, either. I assume everyone is using `https` but I suppose there's a chance someone could add `dev.coder.com` and want to use a less secure protocol.